cli_helpers/tabular_output/output_formatter.py

```diff
@@ -0,0 +1,172 @@
+# -*- coding: utf-8 -*-
+"""A generic tabular data output formatter interface."""
+
+from __future__ import unicode_literals
+from collections import namedtuple
+
+from . import (delimited_output_adapter, vertical_table_adapter,
+               tabulate_adapter, terminaltables_adapter)
+
+MISSING_VALUE = '<null>'
+
+OutputFormatHandler = namedtuple(
+    'OutputFormatHandler',
+    'format_name preprocessors formatter formatter_args')
+
+
+class TabularOutputFormatter(object):
+    """An interface to various tabular data formatting libraries.
+
+    The formatting libraries supported include:
+      - `tabulate <https://bitbucket.org/astanin/python-tabulate>`_
+      - `terminaltables <https://robpol86.github.io/terminaltables/>`_
+      - a CLI Helper vertical table layout
+      - delimited formats (CSV and TSV)
+
+    :param str format_name: An optional, default format name.
+
+    Usage::
+
+      >>> from cli_helpers.tabular_output import TabularOutputFormatter
+      >>> formatter = TabularOutputFormatter(format_name='simple')
+      >>> data = ((1, 87), (2, 80), (3, 79))
+      >>> headers = ('day', 'temperature')
+      >>> print(formatter.format_output(data, headers))
+        day    temperature
+      -----  -------------
+          1             87
+          2             80
+          3             79
+
+    You can use any :term:`iterable` for the data or headers::
+
+      >>> data = enumerate(('87', '80', '79'), 1)
+      >>> print(formatter.format_output(data, headers))
+        day    temperature
+      -----  -------------
+          1             87
+          2             80
+          3             79
+
+    """
+
+    _output_formats = {}
+
+    def __init__(self, format_name=None):
+        """Set the default *format_name*."""
+        self._format_name = format_name
+
+    def set_format_name(self, format_name):
+        """Set the default format.
+
+        :param str format_name: The display format name.
+
+        """
+        if format_name in self.supported_formats():
+            self._format_name = format_name
+        else:
+            raise ValueError('unrecognized format_name: {}'.format(
+                format_name))
+
+    def get_format_name(self):
+        """Get the current default format.
+
+        :return: The format name.
+        :rtype: str or None
+
+        """
+        return self._format_name
+
+    def supported_formats(self):
+        """Return the names of supported output formats.
+
+        :return: The format names.
+        :rtype: tuple
+
+        """
+        return tuple(self._output_formats.keys())
+
+    @classmethod
+    def register_new_formatter(cls, format_name, handler, preprocessors=(),
+                               kwargs={}):
```

Comment:
Hmm. I'm not a Python guru so perhaps I misunderstand the purpose of `kwargs={}` here. I thought mutable default parameters is not a good thing. Does this serve some special purpose?